Improve Testing

cstyan opened the pull request:
First step towards improved testing of the node-adb-client.

So far I've added e2e-tests for all the command types that are supported by the library. Still need to figure out how to use mocks so that code that requires a connection to a device can be unit tested. I copied ContactManager.apk from `appium-adb` to use as a test apk.

Some tests (for example push/pull/install) rely on the results of previous tests (a file being available on the device) which I know is bad, how should I fix that? Should I be using `before` to put the file on the device again? That would fail if the code for `push` is broken which means `push`, `pull`, and `install` tests would all fail anyways.

I've also added a coveralls badge and `gulp coveralls` to the repo (prior to this PR), as well as turned on Coveralls support for this repo, but the coverage report doesn't seem to be updating.

imurchie commented:
Re: push/pull, I would use temporary files so that you can distinguish the file you just pushed from any other. See, for instance, https://github.com/appium/appium-android-driver/blob/master/test/functional/commands/file-movement-e2e-specs.js

imurchie commented:
The easiest way to use mocks is the `appium-test-support` module.

See https://github.com/appium/appium-android-driver/blob/master/test/unit/commands/touch-specs.js

imurchie commented on package.json:
```diff
@@ -45,6 +45,7 @@
     "gulp": "^3.8.11",
     "mochawait": "^2.0.0",
     "proxyquire": "^1.6.0",
+    "sleep": "^3.0.1",
```
Both `asyncbox` and `bluebird` have "sleep" functionality. To wait for a second:

```js
import B from 'bluebird';

await B.delay(1000);
```

```js
import { sleep } from 'asyncbox';

await sleep(1000);
```

cstyan commented:
Hey @imurchie, thanks for the review! I think I've addressed your comments in the most recent commits. I've also added some more unit tests for functions that were, to me, easily testable via mocks. There are more functions and case I'd like to test but this should get the bulk out of the way. Let me know what you think.

I'm still having trouble getting the coverage report to generate. Normally I wouldn't care, but part of what I committed to in my project proposal was getting 70% test coverage.

imurchie commented:
Interesting. Looking around, it doesn't seem like coverage has been sent for a few months now, for any of our packages. I'll have to dig into this.

imurchie commented on lib/usb-device.js:
```diff
@@ -16,7 +16,7 @@ class USBDevice {
 
   // opens the device and sets up the endpoints for out object
-  async claimDevice () {
+  claimDevice () {
```
I don't know how public this method is, but I would tend to make everything in an exposed API, and most other stuff, `async` even if it doesn't need it. It makes it easier to know how to call the functions, and cost basically nil.

cstyan replied:
It should only be called by `adb-device`'s `initConnection` function, so not public at all. The function doesn't need to `await` on any other calls so I removed the `async`.  What's the reasoning for making everything `async` even if they don't need to be, just for consistency's sake I guess?

imurchie replied:
Yeah. Bugs from not waiting for async functions are hard to debug, so just making all public calls have it, even if they don't do any promise work, helps alleviate some of that. That's what we've been doing, anyway. Internal functions don't need to be that way, if you trust the folks working on it (i.e., you!).

cstyan replied:
Good to know. I'll do a read through of all my public facing functions and make sure things are consistent. Once I get testing up to an acceptable level my next todo is documenting all public facing functions for people who want to use the library, as well as internal functions that would need to be implemented for a TCP Device class. 

I'll probably end up writing more documentation than anyone will ever need :)

imurchie commented on test/functional/adb-e2e-specs.js:
```diff
@@ -0,0 +1,236 @@
+  });
+  describe('push tests', () => {
+    const smallFile = path.resolve(__dirname, '..', '..', '..', 'test', 'fixtures', 'smallFile');
+    const largeFile = path.resolve(__dirname, '..', '..', '..', 'test', 'fixtures', 'largeFile');
```
Not sure how large `largeFile` is, but it might be worthwhile to just build a temporary file instead of having a huge fixture.

cstyan replied:
It's current 191k, it needs to be at least larger than 64k in order to test the protocols handling files properly. I'll look at generating a temp. file instead of cloning this file all over the place.

imurchie commented on test/functional/adb-e2e-specs.js:
```diff
@@ -0,0 +1,236 @@
+      let stats = await fs.stat(destination);
+      fileSize.should.equal(stats.size);
+    });
+    it('should return a filesize of -1 if the file doesn\'t exist', async () => {
```
I personally just like to avoid contractions in messages. ;)

cstyan replied:
Force of habit :) changes them all to `does not`

imurchie commented:
A couple of small thoughts. But it looks great in general!

cstyan commented:
@imurchie Added a bunch more tests. Essentially the only functions that don't have tests now are constructors and `usb.claimDevice`. There's one function I realized isn't even been used as I was writing tests for it, I'll remove that function.

Couldn't get the creation of a temporary file to work nicely so I reduced the size of `largeFile` to just over what I need it to be in order to test test the protocols 64k chunking limitation. If cloning around a 71k file is still a concern then I can give the temp. file another go.

If you don't have any major concerns by the end of the week then I'll merge. Next up will be documentation and updated usage examples.

imurchie commented:
I think a 71k file is fine to be in the repo. :)

imurchie commented on test/functional/adb-e2e-specs.js:
```diff
@@ -0,0 +1,248 @@
+process.env.NODE_ENV = 'test';
+
+// const should = chai.should();
+// const expect = chai.expect;
```
Any reason to keep these in?

imurchie commented on test/unit/adb-device-specs.js:
```diff
@@ -0,0 +1,169 @@
+
+// let fsStub = { '@noCallThru': true };
+// let ADBDevice = proxyquire('../../lib/adb-device', { 'fs': fs });
+// fsStub.stat = () => { return "nothing"; };
```
Are these needed?

imurchie commented:
Looks good to me, other than some commented out code that seems unnecessary.

imurchie commented:
Also: are you going to squash it into more logical commits?

cstyan commented:
Removed commented out code. I'll go through the history and see which commits make sense to squash.

I'm curious what you think about the tests for the connection state machine: https://github.com/appium/node-adb-client/blob/callum-improve-testing/test/unit/adb-specs.js#L176-L285

Also is it okay to have multiple `should` conditions such as in this test: https://github.com/appium/node-adb-client/blob/callum-improve-testing/test/unit/helpers-specs.js#L75-L99

imurchie commented:
Multiple `should` statements seems fine in that case. You're just checking that all the different data bits are not there. It would be weird to do it in different tests.

The connection state machine tests look good to me. Are all the permutations of cases accounted for? If so, good to go!

cstyan commented:
I _think_ I got all the cases, but I'll go through the state machine again tonight and see if I missed any.

cstyan commented:
Hey @imurchie, I think this is done. I've rebased and squashed some of the commits, wasn't able to squash anymore due to the merge commit in the middle of the branch history (I think). 

Just waiting for your 👍 
Should we just squash all commits when merging the PR?

imurchie commented:
In general it is best to squash into the logical number of commits. That is, let each commit add something specific. It's not always possible and if the PR is older it will get harder and harder.

Code-wise, this is awesome! 👍

cstyan commented:
I'm fairly certain it's the merge commit that's causing issues when I try to squash any further. Initially I didn't know you could reorder commits when rebasing either. In the future I will proactively squash as I work so we don't run into this issue again.

Merging, I will update the examples and add documentation regarding the classes/functions tonight.

imurchie commented:
Yeah. Merge commits. Bleh.
